runtime/memory/lazy.c

```diff
@@ -0,0 +1,143 @@
+#include "../runtime.h"
+
+#ifndef WASM_BLOCK_SIZE
+#define WASM_BLOCK_SIZE 4096
+#endif
+
+void **memory_blocks;
+u32 memory_size;
+
+void alloc_linear_memory() {
+    memory_blocks = malloc((starting_pages * WASM_PAGE_SIZE / WASM_BLOCK_SIZE)
+            * sizeof(void*));
+    memset(memory_blocks, 0, (starting_pages * WASM_PAGE_SIZE / WASM_BLOCK_SIZE)
+            * sizeof(void*));
+    memory_size = starting_pages * WASM_PAGE_SIZE;
+    printf("mem size: %d\n", memory_size);
+}
+
+void expand_memory() {
+    printf("mem size: %d -> %d\n", memory_size, memory_size + WASM_PAGE_SIZE);
+    // max_pages = 0 => no limit
+    assert(max_pages == 0 || (memory_size / WASM_PAGE_SIZE < max_pages));
+
+    memory_blocks = realloc(memory_blocks,
+            ((memory_size + WASM_PAGE_SIZE) / WASM_BLOCK_SIZE) * sizeof(void*));
+    assert(memory_blocks);
+
+    memory_blocks[memory_size / WASM_BLOCK_SIZE] = NULL;
+    memory_size += WASM_PAGE_SIZE;
+}
+
+// FIXME: This can't be provided by this memory backend
+//INLINE char* get_memory_ptr_for_runtime(u32 offset, u32 bounds_check) {
+//    assert(memory_size > bounds_check && offset <= memory_size - bounds_check);
+//
+//    char* mem_as_chars = (char *) memory;
+//    char* address = &mem_as_chars[offset];
+//
+//    return address;
+//}
+
+// All of these are pretty generic
+INLINE i8 get_i8(i32 offset) {
+    assert(offset <= memory_size - sizeof(i8));
+
+    void *block = memory_blocks[offset / WASM_BLOCK_SIZE];
+    if (!block) {
+        return 0;
+    }
+
+    return ((i8*)block)[offset % WASM_BLOCK_SIZE];
+}
+
+INLINE i16 get_i16(i32 offset) {
+    return ((u16)(u8)get_i8(offset+0) << 0) |
+           ((u16)(u8)get_i8(offset+1) << 8);
+}
+
+INLINE i32 get_i32(i32 offset) {
+    return ((u32)(u8)get_i8(offset+0) << 0) |
+           ((u32)(u8)get_i8(offset+1) << 8) |
+           ((u32)(u8)get_i8(offset+2) << 16) |
+           ((u32)(u8)get_i8(offset+3) << 24);
+}
+
+INLINE i64 get_i64(i32 offset) {
+    return ((u64)(u8)get_i8(offset+0) << 0) |
+           ((u64)(u8)get_i8(offset+1) << 8) |
+           ((u64)(u8)get_i8(offset+2) << 16) |
+           ((u64)(u8)get_i8(offset+3) << 24) |
+           ((u64)(u8)get_i8(offset+4) << 32) |
+           ((u64)(u8)get_i8(offset+5) << 40) |
+           ((u64)(u8)get_i8(offset+6) << 48) |
+           ((u64)(u8)get_i8(offset+7) << 56);
+}
```

Comment:
Can you lift the bounds checks out here? No need to check the assert so many times (same for all the other repeated loads)